From f8e91a533342fcf82a54e9ef376b1161f8740768 Mon Sep 17 00:00:00 2001
From: Brandon Hurst <brandon.hurst@analog.com>
Date: Tue, 17 Feb 2026 10:56:47 -0800
Subject: [PATCH 34/39] drivers: usb: udc: max32: Handle deferred endpoint
 transfers

Issue was corrected where transfers waiting on a busy
endpoint were not being serviced. This was causing some transfers
to never complete.

Signed-off-by: Brandon Hurst <brandon.hurst@analog.com>
---
 drivers/usb/udc/udc_max32.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/usb/udc/udc_max32.c b/drivers/usb/udc/udc_max32.c
index 9897cc179a9..e2bb673c2e8 100644
--- a/drivers/usb/udc/udc_max32.c
+++ b/drivers/usb/udc/udc_max32.c
@@ -233,6 +233,11 @@ static void udc_event_xfer_in_done(const struct device *dev, struct udc_ep_confi
 	} else {
 		udc_submit_ep_event(dev, buf, 0);
 	}
+
+	/* Start the next transfer if there is another waiting */
+	if (ep_cfg->addr != USB_CONTROL_EP_IN && udc_buf_peek(ep_cfg) != NULL) {
+		udc_event_xfer_in(dev, ep_cfg);
+	}
 }
 
 static void udc_event_xfer_out_done(const struct device *dev, struct udc_ep_config *ep_cfg)
@@ -260,6 +265,11 @@ static void udc_event_xfer_out_done(const struct device *dev, struct udc_ep_conf
 	} else {
 		udc_submit_ep_event(dev, buf, 0);
 	}
+
+	/* Start the next transfer if there is another waiting */
+	if (ep_cfg->addr != USB_CONTROL_EP_OUT && udc_buf_peek(ep_cfg) != NULL) {
+		udc_event_xfer_out(dev, ep_cfg);
+	}
 }
 
 static int udc_ctrl_feed_dout(const struct device *dev, const size_t length)
-- 
2.47.3

